lib/src/http.dart

```diff
@@ -38,14 +40,24 @@ final _sessionId = createUuid();
 /// user-friendly error messages.
 class _PubHttpClient extends http.BaseClient {
   final _requestStopwatches = <http.BaseRequest, Stopwatch>{};
+  SystemCache get cache => _cache ??= SystemCache();
+  SystemCache _cache;
 
   http.Client _inner;
 
   _PubHttpClient([http.Client inner]) : _inner = inner ?? http.Client();
 
   @override
   Future<http.StreamedResponse> send(http.BaseRequest request) async {
-    if (_shouldAddMetadata(request)) {
+    // check for a token in case of hosted service
+    var token = getToken(cache, request.url);
+    if (token != null) {
+      if (!request.headers.containsKey(HttpHeaders.authorizationHeader)) {
+        request.headers[HttpHeaders.authorizationHeader] = 'Bearer $token';
```

Comment:
We need to determine if this should be stripped when following redirects, and whether or not it is.

I haven't thought this through yet, but we need to look out for vulnerabilities where one server redirects to another...

Reply:
See my comment above.

Reply:
`var token = getToken(cache, request.url);`, yeah okay, auth won't be added to requests when they are redirected.

I guess, we should think about whether or not it's reasonable that authentication may be added to the `archieve_url` returned when listing versions.

https://github.com/dart-lang/pub/blob/73f0906e152a2de98ec881523f289db04f5b3b9b/lib/src/source/hosted.dart#L371-L376

I think my concern here is that server-A could return a response for package `foo` which tricks the client into making a `GET` request against any resource on server-B with credentials for server-B.

I guess we should consider in:
https://github.com/dart-lang/pub/blob/73f0906e152a2de98ec881523f289db04f5b3b9b/lib/src/source/hosted.dart#L399

That it's only allowed to use credentials to make the request, if `versionInfo.archiveUrl` belongs to `server = source._parseDescription(id.description).last`.

Otherwise, a malicious pub server could return an archive URL and tricks the client into making an authenticated GET request against a resource of the attackers choosing.

Reply:
Okay I agree. Basically, if the archive_url is not the same origin, no tokens are added, and it's the third-party responsibility to create a one-time secure url to download the package.

Reply:
All tested and working. Third party servers can have a path and current code fixed. Left uploader code as it is because it is not supported for third party servers.
No possibility for tricks; tokens are only added if archive_url is from same path where token exists for.